---
title: "How to use KNN sleepwalk on single-cell sequencing data"
output: html_document
date: "2023-01-25"
---

This code tests the wrapper function for a KNN-based sleepwalk, particularly for single cell data. We use the dataset and [Seurat vignette](https://satijalab.org/seurat/articles/pbmc3k_tutorial.html) here as a template. 

```{r functions}
source('src/knn_sleepwalk.R')
set.seed(1234)
```

```{r process_data}
library(dplyr)
library(Seurat)
library(patchwork)

# Load the PBMC dataset
pbmc.data <- Read10X(data.dir = "data/filtered_gene_bc_matrices/hg19/")
# Initialize the Seurat object with the raw (non-normalized data).
pbmc <- CreateSeuratObject(counts = pbmc.data, project = "pbmc3k", min.cells = 3, min.features = 200)

# Processing
pbmc[["percent.mt"]] <- PercentageFeatureSet(pbmc, pattern = "^MT-")
pbmc <- NormalizeData(pbmc, normalization.method = "LogNormalize", scale.factor = 10000)
pbmc <- FindVariableFeatures(pbmc, selection.method = "vst", nfeatures = 2000)
top10 <- head(VariableFeatures(pbmc), 10)
all.genes <- rownames(pbmc)
pbmc <- ScaleData(pbmc, features = all.genes)
pbmc <- RunPCA(pbmc, features = VariableFeatures(object = pbmc))
```

                                            
```{r run_dimr}
# Dimr
pbmc <- RunUMAP(pbmc, dims = 1:10)

# Tidy
umap <- pbmc@reductions$umap@cell.embeddings %>% as_tibble()

# And add pca as the original manifold
pca <- pbmc@reductions$pca@cell.embeddings %>% as_tibble()
var_genes <- pbmc@assays$RNA@var.features
hid <- pbmc@assays$RNA@scale.data[var_genes,] %>% t() %>% as_tibble()
```

```{r run_knn_sleepwalk}
setwd(here::here("output"))
KnnSleepwalkDirect(mat1 = umap, mat2 = pca, embedding = list(umap, umap), k = 27, point_size = 4, output_file = "knn_sleepwalk_scrna_seq_umap_vs_pca_space.html")
```
Now we do this for K-farthest neighbors.

```{r run_kfn_sleepwalk}
setwd(here::here("output"))
KfnSleepwalkDirect(mat1 = umap, mat2 = pca, embedding = list(umap, umap), k = 270, point_size = 4, output_file = "kfn_sleepwalk_scrna_seq_umap_vs_pca_space.html")
```
This is very difficult to interpret. So we're going to try a few tricks. Let's just run regular sleepwalk with the distance in UMAP space versus regular space. Then we invert it, where it is slightly easier to see that far away things in umap space are not equal to far away things in pca space.


```{r}
setwd(here::here("output"))
umap_dist <- dist(umap) %>% as.matrix()
pca_dist <- dist(pca) %>% as.matrix()

# Normal
sleepwalk::sleepwalk(embeddings = list(umap, umap), 
                     compare = "distances", 
                     distances = list(umap_dist, pca_dist), 
                     saveToFile = "sleepwalk_scrna_seq_umap_vs_pca.html",
                     pointSize = 4)

# Inverted
sleepwalk::sleepwalk(embeddings = list(umap, umap), 
                     compare = "distances", 
                     distances = list(1/umap_dist, 1/pca_dist), 
                     saveToFile = "sleepwalk_inverted_scrna_seq_umap_vs_pca.html",
                     pointSize = 4)
```

What we see is that in the inverted case, there's all kinds of information there that UMAP sweeps under the rug. Like a particular region that seems to be apart from everyone.



